### DISCUSSION TITLE
Flex with expanding spacer

### DISCUSSION THREAD

**Discussion by abau** [fltk-rs/fltk-rs#1223] [Q&A] [answered]
I need to display dynamic data in a row where there is a left side (containing some data), an expanding empty spacer and a right side (containing some other data):
```
data 1    data 2  ~~~~ expanding spacer ~~~~  data 4    data 5
```
In this demo, there is a button dynamically adding frames each containing a string to be displayed:
```
use fltk::app::App;
use fltk::button::Button;
use fltk::frame::Frame;
use fltk::group::{Column, Flex};
use fltk::prelude::*;
use fltk::window::Window;

fn main() {
    let app = App::default();
    let mut window = Window::default().with_size(800, 200).center_screen();
    let mut data: Vec<String> = Default::default();
    let column = Column::default_fill();
    let mut button = Button::new(0, 0, 800, 100, "Add data");
    let mut flex = Flex::new(0, 100, 800, 100, "").row();

    flex.end();
    column.end();
    window.end();
    window.show();

    button.set_callback(move |_| {
        data.push(data.len().to_string());

        flex.clear();

        // Left
        for d in &data {
            let frame = Frame::default().with_label(&d);

            flex.add(&frame);
            flex.set_size(&frame, 20);
        }

        // Spacer
        flex.add(&Frame::default());

        // Right
        for d in &data {
            let frame = Frame::default().with_label(&d);

            flex.add(&frame);
            flex.set_size(&frame, 20);
        }

        flex.recalc();
        flex.redraw();
    });

    app.run().unwrap();
}
```
Initially:
![1](https://user-images.githubusercontent.com/4090497/175386390-8952fccf-886e-4da8-bd47-4bdc86530a83.png)

After clicking "Add data" once, the GUI behaves as expected:

![2](https://user-images.githubusercontent.com/4090497/175386397-5c9b3626-9f22-4cc9-9852-64b85b339513.png)

But after the second click on "Add data", the spacer does not seem to be taken into account when laying out the frames:

![3](https://user-images.githubusercontent.com/4090497/175386400-8fde9c39-5263-498d-a480-baca82bf3299.png)

This continues when continually clicking "Add data":

![4](https://user-images.githubusercontent.com/4090497/175386402-19cae57d-7955-4c21-bd10-2f8bde7abf8b.png)

What am I doing wrong here?

**Comment by MoAlyousef** [comment, marked as answer]
I'm not sure what the issue is, I'll have to dig deeper. But it seems the empty zero-sized frame messes up the flex dynamic calculation.
There is a workaround:
```rust
use fltk::app::App;
use fltk::button::Button;
use fltk::enums::{Color, FrameType};
use fltk::frame::Frame;
use fltk::group::{Column, Flex};
use fltk::prelude::*;
use fltk::window::Window;

fn main() {
    let app = App::default();
    let mut window = Window::default().with_size(800, 200).center_screen();
    let mut data: Vec<String> = Default::default();
    let column = Column::default_fill();
    let mut button = Button::new(0, 0, 800, 100, "Add data");
    let mut flex = Flex::new(0, 100, 800, 100, "").row();

    flex.end();
    column.end();
    window.end();
    window.show();

    button.set_callback(move |_| {
        data.push(data.len().to_string());

        flex.clear();

        // Left
        for d in &data {
            let frame = Frame::default().with_label(&d);

            flex.add(&frame);
            flex.set_size(&frame, 20);
        }

        // Spacer
        flex.add(
            &Frame::default().with_size(flex.w() - data.len() as i32 * (20 + flex.pad()) * 2, 0),
        );

        // Right
        for d in &data {
            let frame = Frame::default().with_label(&d);

            flex.add(&frame);
            flex.set_size(&frame, 20);
        }

        flex.recalc();
        flex.redraw();
    });

    app.run().unwrap();
}
```

**Reply by abau** [reply]
> There is a workaround:

Unfortunately, my real-life use-case is more complex than this demo, so I hoped I could get around of manually computing the spacers width.

> I'm not sure what the issue is, I'll have to dig deeper. But it seems the empty zero-sized frame messes up the flex dynamic calculation.

Thanks for your effort. 

<details>
<summary>I've modified the demo so that initially, `data` is not empty (click here!)</summary>

```
use fltk::app::App;
use fltk::button::Button;
use fltk::enums::{Color, FrameType};
use fltk::frame::Frame;
use fltk::group::{Column, Flex};
use fltk::prelude::*;
use fltk::window::Window;

fn update(data: &mut Vec<String>, flex: &mut Flex) {
    data.push(data.len().to_string());

    flex.clear();

    // Left
    for d in data.iter() {
        let frame = Frame::default().with_label(&d);

        flex.add(&frame);
        flex.set_size(&frame, 20);
    }

    // Spacer
    flex.add(
        &Frame::default().with_size(flex.w() - data.len() as i32 * (20 + flex.pad()) * 2, 0),
    );

    // Right
    for d in data.iter() {
        let frame = Frame::default().with_label(&d);

        flex.add(&frame);
        flex.set_size(&frame, 20);
    }

    flex.recalc();
    flex.redraw();
}

fn main() {
    let app = App::default();
    let mut window = Window::default().with_size(800, 200).center_screen();
    let mut data = vec!["0".to_string(), "1".to_string()];
    let column = Column::default_fill();
    let mut button = Button::new(0, 0, 800, 100, "Add data");
    let mut flex = Flex::new(0, 100, 800, 100, "").row();

    flex.end();
    column.end();
    window.end();
    window.show();

    update(&mut data, &mut flex);

    button.set_callback(move |_| {
        update(&mut data, &mut flex);
    });

    app.run().unwrap();
}
```  
</details>

I observed two things:

* It seems like adding to the flex for the first time always works fine, but occasionally fails after `clear`ing after widgets have already been added
* When running the code from my original post, I cannot reproduce the exact behavior I was able to reliably reproduce two days ago. Still, it fails, but the pattern is different.

Thus, maybe I'm not correctly "destroying" previous children of the flex (not sure how the underlying C++ FLTK lib handles them) and they still are referenced by the flex in some undeterministic way?

**Reply by MoAlyousef** [reply]
Yeah the Fl_Flex::clear() method wasn't implemented, so it used Fl_Group::clear() which had no idea of Fl_Flex's internal vec. I've pushed a fix on master if you can try it:
```toml
[dependencies]
fltk = { version = "1.3.10", git="https://github.com/fltk-rs/fltk-rs" }
```

**Reply by abau** [reply]
Appears to work well for the demo above as well as my real-world application.

Thank you very much!
